File: app/src/main/java/oblitusnumen/calendar/implementation/data/Date.java
package oblitusnumen.calendar.implementation.data;


import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.provider.BaseColumns;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.ArrayList;

public class Date implements BaseColumns {
    public static final String TABLE_NAME = "dates";
    public static final String COLUMN_NAME_ID = "id";
    public static final String COLUMN_NAME_ENTRY_ID = "entryId";
    public static final String COLUMN_NAME_DESC = "description";
    public static final String COLUMN_NAME_TIME_START = "timeStart";
    public static final String COLUMN_NAME_DURATION = "duration";
    public static final String COLUMN_NAME_TIME_ENDS = "timeEnd";
    public static final String COLUMN_NAME_TIMES_REPEATS = "timesRepeat";
    public static final String COLUMN_NAME_PERIOD = "period";
    public static final String COLUMN_NAME_TIME_ZONE = "timeZone";
    public static final String COLUMN_NAME_REMOVED = "exceptionRules";
    private final DbManager dbManager;
    int id = -1;
    int entryId;
    String desc;
    long start;
    long duration;
    long end;
    int timesRepeat = 1;
    Period period;
    ZoneId zoneId;
    ExceptionRules exceptionRules;

    Date(DbManager dbManager, Cursor cursor) {
        this(dbManager, cursor.getInt(cursor.getColumnIndex(COLUMN_NAME_ID)),
                cursor.getInt(cursor.getColumnIndex(COLUMN_NAME_ENTRY_ID)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_DESC)),
                cursor.getLong(cursor.getColumnIndex(COLUMN_NAME_TIME_START)),
                cursor.getLong(cursor.getColumnIndex(COLUMN_NAME_DURATION)),
                cursor.getLong(cursor.getColumnIndex(COLUMN_NAME_TIME_ENDS)),
                cursor.getInt(cursor.getColumnIndex(COLUMN_NAME_TIMES_REPEATS)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_PERIOD)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_TIME_ZONE)),
                cursor.getString(cursor.getColumnIndex(COLUMN_NAME_REMOVED)));
    }

    Date(DbManager dbManager, int id, int entryId, String desc, long start, long duration, long end, int timesRepeat, String period, String zoneId, String removed) {
        this.dbManager = dbManager;
        this.id = id;
        this.entryId = entryId;
        this.desc = desc;
        this.start = start;
        this.duration = duration;
        this.end = end;
        setupTimesRepeat(timesRepeat);
        this.period = new Period(period);
        this.zoneId = ZoneId.of(zoneId);
        this.exceptionRules = new ExceptionRules(this, removed);
    }

    public Date(DbManager dbManager, Entry entry, String desc, ZonedDateTime time, long duration, int timesRepeat, Period period) {
        this.dbManager = dbManager;
        this.entryId = entry.getId();
        this.desc = desc;
        this.start = time.toEpochSecond();
        this.duration = duration;
        this.period = period;
        this.zoneId = time.getZone();
        this.exceptionRules = ExceptionRules.none(this);
        setTimesRepeat(timesRepeat);
    }

    void createOrUpdate() {
        if (!isEmpty()) {
            ContentValues contentValues = toContentValues();
            contentValues.put(COLUMN_NAME_ID, (Integer) null);
            id = (int) dbManager.getWritableDatabase().insertWithOnConflict(TABLE_NAME, null, contentValues, SQLiteDatabase.CONFLICT_REPLACE);
        } else {
            delete();
        }
    }

    public boolean isEndless() {
        return timesRepeat == Integer.MAX_VALUE;
    }

    public boolean isEmpty() {
        return timesRepeat <= 0;
    }

    public void delete() {
        dbManager.getWritableDatabase().execSQL("DELETE FROM " + TABLE_NAME + " WHERE " + COLUMN_NAME_ID + " = ?", new String[]{String.valueOf(id)});

        // FIXME: 10/24/24 remove all associated entries i.e.
    }

    ContentValues toContentValues() {
        ContentValues contentValues = new ContentValues();
        contentValues.put(COLUMN_NAME_ID, id);
        contentValues.put(COLUMN_NAME_ENTRY_ID, entryId);
        contentValues.put(COLUMN_NAME_DESC, desc);
        contentValues.put(COLUMN_NAME_TIME_START, start);
        contentValues.put(COLUMN_NAME_DURATION, duration);
        contentValues.put(COLUMN_NAME_TIME_ENDS, end);
        contentValues.put(COLUMN_NAME_TIMES_REPEATS, timesRepeat == Integer.MAX_VALUE ? -1 : timesRepeat);
        contentValues.put(COLUMN_NAME_PERIOD, period.toString());
        contentValues.put(COLUMN_NAME_TIME_ZONE, zoneId.toString());
        contentValues.put(COLUMN_NAME_REMOVED, exceptionRules.toString());
        return contentValues;
    }

    /**
     * @return null if event does not happen during next day from <code>startOfDay</code> or time which event takes place at
     */
    public ZonedDateTime forDay(ZonedDateTime startOfDay) {
        long start = startOfDay.toEpochSecond();
        long end = startOfDay.plusDays(1).toEpochSecond();
        int zonedDateTimeIndex = getZonedDateTimeIndex(start, end);
        if (zonedDateTimeIndex == -1 || !exceptionRules.isEventPresent(zonedDateTimeIndex)) return null;
        return getZoneDateTime(zonedDateTimeIndex);
    }

    public ZonedDateTime getZoneDateTime(ZoneId zoneId, int idx) {
        return getZoneDateTime(idx).withZoneSameInstant(zoneId);
    }

    public ZonedDateTime getZoneDateTime(int idx) {
        return period.getTime(Instant.ofEpochSecond(start).atZone(zoneId), idx);
    }

    public Period getPeriod() {
        return period;
    }

    long getTime(int idx) {
        return getZoneDateTime(idx).toEpochSecond();
    }

    public String getDesc() {
        return desc.isEmpty() ? getEntry().getName() : desc;
    }

    public boolean exists() {
        return id != -1;
    }

    public void setDesc(String desc) {
        this.desc = desc;
    }

    public Entry getEntry() {
        return Entry.Companion.byId(dbManager, entryId);
    }

    public long getStart() {
        return start;
    }

    private void setupTimesRepeat(int timesRepeat) {
        this.timesRepeat = timesRepeat == -1 ? Integer.MAX_VALUE : timesRepeat;
    }

    /**
     * @return -1 if date is endless otherwise count of date occurring (without respect to deleted occurrences)
     */
    public int getTimesRepeat() {
        return isEndless() ? -1 : timesRepeat;
    }

    private void setTimesRepeat(int timesRepeat) {
        this.timesRepeat = timesRepeat;
        this.end = getTime(timesRepeat - 1);
    }

    int getZonedDateTimeIndex(long start, long finish) {
        if (finish <= this.start) return -1;
        if (this.end == this.start) return this.start >= start ? 0 : -1;
        long period = (this.end - this.start) / timesRepeat;
        int idx = Math.min(timesRepeat - 1, (int) ((finish - this.start) / period));
        long time = getZoneDateTime(idx).toEpochSecond();
        if (time >= start && time < finish)
            return idx;
        if (time >= finish && idx > 1) {
            long timem1 = getZoneDateTime(idx - 1).toEpochSecond();
            if (timem1 >= finish || timem1 < start)
                return -1;
            else
                return idx - 1;
        }
        if (time < start && idx < timesRepeat - 1) {
            long timep1 = getZoneDateTime(idx + 1).toEpochSecond();
            if (timep1 < start || timep1 >= finish)
                return -1;
            else
                return idx + 1;
        }
        return -1;
    }

    void addEvent(int idx) {
        exceptionRules.addEvents(idx, idx + 1);
    }

    void addEvents(int from, int to) {
        exceptionRules.addEvents(from, to);
    }

    void makeEndless() {
        exceptionRules.addEvents(timesRepeat, Integer.MAX_VALUE);
    }

    void cropToTimesRepeat(int timesRepeat) {
        if (timesRepeat > this.timesRepeat) {
            addEvents(this.timesRepeat, timesRepeat);
        } else exceptionRules.removeEvents(timesRepeat, Integer.MAX_VALUE);
    }

    void removeEvent(int idx) {
        exceptionRules.removeEvents(idx, idx + 1);
    }

    void removeEvents(int from, int to) {
        exceptionRules.removeEvents(from, to);
    }

    static class ExceptionRules {
        final ArrayList<Rule> rules = new ArrayList<>();
        private final Date date;

        ExceptionRules(Date date, String removed) {
            this.date = date;
            if (removed.isEmpty()) return;
            for (String s : removed.split(",")) {
                rules.add(new Rule(s));
            }
        }

        static ExceptionRules none(Date date) {
            return new ExceptionRules(date, "");
        }

        private void clearRules(int from, int to) {
            rules.subList(from, to).clear();
        }

        private void addEvents(int from, int to) {//end not including
            if (from > date.timesRepeat) {//this does not create any cases like 5-10,10-20, as last event in date must occur
                rules.add(new Rule(date.timesRepeat, from));
            }
            int idx0 = findIndex(from);//index after rule to be changed before interval
            if (to > date.timesRepeat) {//crop date to size to
                date.setTimesRepeat(to);
                if (idx0 > 0) {
                    Rule rule = rules.get(idx0 - 1);
                    if (rule.end > from) {
                        rule.end = from;
                    }
                }
                clearRules(idx0, rules.size());
                return;
            }
            int idx1 = findIndex(to);//index after rule to be changed after interval
            if (idx0 > 0) {//before interval
                Rule rule0 = rules.get(idx0 - 1);
                if (rule0.start == from) {//rule begins from interval
                    idx0--;
                } else {
                    if (rule0.end > to) {//splitting rule
                        rules.add(idx0, new Rule(to, rule0.end));
                        rule0.end = from;
                        return;
                    }//lowering ceiling of rule
                    if (rule0.end > from) {
                        rule0.end = from;
                    }
                }
            }
            if (idx1 > 0) {//index-1 is in range // after interval
                Rule rule1 = rules.get(idx1 - 1);
                if (rule1.end > to) {//rule is not fully inside interval
                    rule1.start = to;
                    idx1--;
                }
            }
            clearRules(idx0, idx1);
        }

        private void removeEvents(int from, int to) {
            if (from >= date.timesRepeat) return;
            int idx0 = findIndex(from);//index after rule to be changed before interval
            if (to >= date.timesRepeat) {//crop date to size from
                if (idx0 > 0) {
                    Rule rule = rules.get(idx0 - 1);
                    if (rule.end >= from) {
                        from = rule.start;
                        idx0--;
                    }
                }
                clearRules(idx0, rules.size());
                date.setTimesRepeat(from);
                return;
            }
            if (idx0 > 0) {//before interval
                Rule rule0 = rules.get(idx0 - 1);
                if (rule0.end >= from) {//end of rule is in interval
                    from = rule0.start;
                    idx0--;
                }
            }
            int idx1 = findIndex(to);//index after rule to be changed after interval
            if (idx1 > 0) {//index-1 is in range // after interval
                Rule rule1 = rules.get(idx1 - 1);
                if (rule1.end > to) {//rule ends after to
                    to = rule1.end;
                }
            }
            clearRules(idx0, idx1);
            rules.add(idx0, new Rule(from, to));
        }

        private int findIndex(int beginIdx) {
            int begin = 0;
            int end = rules.size();
            while (true) {//search algorithm
                int center = (begin + end) / 2;
                if (center == end) return center;
                if (rules.get(center).start > beginIdx) {
                    end = center;
                } else {
                    if (begin == center) return end;
                    begin = center;
                }
            }
        }

        boolean isEventPresent(int idx) {
            int index = findIndex(idx);//for faster work
            if (index > 0) {//may be in some rules
                return rules.get(index - 1).isPresent(idx);
            }
            return true;
        }

        @Override
        public @NotNull String toString() {
            StringBuilder result = new StringBuilder();
            for (Rule rule : rules) {
                result.append(rule).append(",");
            }
            return result.toString();
        }

        private class Rule {
            int start;
            int end;

            Rule(String rule) {
                String[] split = rule.split("-");
                switch (split.length) {
                    case 1 -> {
                        start = Integer.parseInt(split[0]);
                        end = Integer.MAX_VALUE;
                    }
                    case 2 -> {
                        start = Integer.parseInt(split[0]);
                        end = Integer.parseInt(split[1]);
                    }
                    default -> throw new RuntimeException("impossible outcome");
                }
            }

            Rule(int start, int end) {
                this.start = start;
                this.end = end;
            }

            boolean isPresent(int idx) {
                return idx < start || end <= idx;
            }

            @Override
            public @NotNull String toString() {
                return start + "-" + (end == date.timesRepeat ? "" : end);
            }
        }
    }
}

File: app/src/main/java/oblitusnumen/calendar/implementation/data/Date.kt
package oblitusnumen.calendar.implementation.data

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.provider.BaseColumns
import oblitusnumen.calendar.implementation.mult_frac
import oblitusnumen.calendar.implementation.toWeekNumber
import org.jetbrains.annotations.TestOnly
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZonedDateTime
import kotlin.math.min

class Date : BaseColumns {
    private val dbManager: DbManager?
    var id: Int? = null
        private set
    internal var entryId: Int?
    private var desc: String
    var start: Long
        private set
    private var duration: Long
    private var end: Long = 0
    private var timesRepeat: Long = 1
    var period: Period
        private set
    var zoneId: ZoneId
        private set
    var exceptionRules: ExceptionRules
        private set

    val entry: Entry
        get() = dbManager!!.getEntryById(entryId!!)!!

    internal constructor(
        dbManager: DbManager?,
        id: Int,
        entryId: Int,
        desc: String,
        start: Long,
        duration: Long,
        end: Long,
        timesRepeat: Long,
        period: String,
        zoneId: String?,
        removed: String
    ) {
        this.dbManager = dbManager
        this.id = id
        this.entryId = entryId
        this.desc = desc
        this.start = start
        this.duration = duration
        this.end = end
        this.timesRepeat = timesRepeat
        this.period = Period.decode(period)
        this.zoneId = ZoneId.of(zoneId)
        this.exceptionRules = ExceptionRules(removed)
    }

    constructor(
        dbManager: DbManager,
        entry: Entry,
        desc: String,
        time: ZonedDateTime,
        duration: Long,
        timesRepeat: Long,
        period: Period
    ) {
        this.dbManager = dbManager
        this.entryId = entry.id
        this.desc = desc
        this.start = time.toEpochSecond()
        this.duration = duration
        this.period = period
        this.zoneId = time.zone
        this.exceptionRules = ExceptionRules()
        setTimesRepeat(timesRepeat)
    }

    fun create() {
        fixDateRange()
        fixExceptionList()
        if (isEmpty)
            return
        val contentValues = getContentValues()
        contentValues.put(COLUMN_NAME_ID, null as Int?)
        id = dbManager!!.writableDatabase.insertWithOnConflict(
            TABLE_NAME,
            null,
            contentValues,
            SQLiteDatabase.CONFLICT_REPLACE
        ).toInt()
    }

    fun update() {
        fixDateRange()
        fixExceptionList()
        if (isEmpty) {
            delete()
            return
        }
        dbManager!!.writableDatabase.update(
            TABLE_NAME,
            getContentValues(),
            "$COLUMN_NAME_ID = ?",
            arrayOf(id.toString())
        )
    }

    fun delete() {
        dbManager!!.writableDatabase.execSQL(
            "DELETE FROM $TABLE_NAME WHERE $COLUMN_NAME_ID = ?",
            arrayOf(id.toString())
        )
        id = null
    }

    private fun getContentValues(): ContentValues {
        val contentValues = ContentValues()
        contentValues.put(COLUMN_NAME_ENTRY_ID, entryId)
        contentValues.put(COLUMN_NAME_DESC, desc)
        contentValues.put(COLUMN_NAME_TIME_START, start)
        contentValues.put(COLUMN_NAME_DURATION, duration)
        contentValues.put(COLUMN_NAME_TIME_ENDS, end)
        contentValues.put(COLUMN_NAME_TIMES_REPEATS, timesRepeat)
        contentValues.put(COLUMN_NAME_PERIOD, period.toString())
        contentValues.put(COLUMN_NAME_TIME_ZONE, zoneId.toString())
        contentValues.put(COLUMN_NAME_REMOVED, exceptionRules.toString())
        return contentValues
    }

    val isPeriodic: Boolean
        get() = period !is Period.Once

    val isEndless: Boolean
        get() = end > END_ENDLESS_THRESHOLD

    val isEmpty: Boolean
        get() = timesRepeat <= 0

    private fun getZoneDateTime(idx: Long): ZonedDateTime {
        return period.addTo(Instant.ofEpochSecond(start).atZone(zoneId), idx)
    }

    fun getFirstZoneDateTime(): ZonedDateTime {
        return getZoneDateTime(0)
    }

    fun getLastZoneDateTime(): ZonedDateTime {
        if (isPeriodic)
            return getZoneDateTime(timesRepeat - 1)
        return getZoneDateTime(0)
    }

    fun getNext(start: Long): ZonedDateTime? {
        var date: ZonedDateTime? = null
        var curStart = start
        while (date == null) {
            val nextClosestRaw = getNextClosestRaw(curStart) ?: break
            val epochSecond = nextClosestRaw.toEpochSecond()
            if (exceptionRules.getRangeForDate(nextClosestRaw.toEpochDays()) == null) date = nextClosestRaw
            curStart = epochSecond + 1
        }
        return date
    }

    fun getAllInRange(start: Long, end: Long): List<ZonedDateTime> {
        val dates: MutableList<ZonedDateTime> = ArrayList()
        var curStart = start
        while (true) {
            val nextClosestRaw = getNextClosestRaw(curStart) ?: break
            val epochSecond = nextClosestRaw.toEpochSecond()
            if (epochSecond >= end) break
            if (exceptionRules.getRangeForDate(nextClosestRaw.toEpochDays()) == null) dates.add(nextClosestRaw)
            curStart = epochSecond + 1
        }
        return dates
    }

    private fun getZonedDateTimeInRange(start: Long, finish: Long): ZonedDateTime? { //any(?) in range
        if (finish <= this.start || timesRepeat == 0L) return null
        if (this.end == this.start) return if (this.start >= start) getZoneDateTime(0) else null
        val periodExpect = (this.end - this.start) / (timesRepeat - 1)
        val idxExpect = (finish - this.start) / periodExpect
        //mult_frac will overflow with period 1D after a few million years, but who cares about that
        val timeEst = mult_frac(this.end - this.start, idxExpect, timesRepeat) + this.start
        val idxDiff = (timeEst - start) / periodExpect
        val idx = min((timesRepeat - 1), idxExpect - idxDiff)
        val zdtIdx = getZoneDateTime(idx)
        val time = zdtIdx.toEpochSecond()
        if (time in start..<finish) return zdtIdx
        if (time >= finish && idx > 1) {
            val zdtIdxM1 = getZoneDateTime(idx - 1)
            val timeM1 = zdtIdxM1.toEpochSecond()
            return if (timeM1 >= finish || timeM1 < start) null
            else zdtIdxM1
        }
        if (time < start && idx < timesRepeat - 1) {
            val zdtIdxP1 = getZoneDateTime(idx + 1)
            val timeP1 = zdtIdxP1.toEpochSecond()
            return if (timeP1 < start || timeP1 >= finish) null
            else zdtIdxP1
        }
        return null
    }

    /**
     * @return null if event does not happen during next day from `startOfDay` or time which event takes place at
     */
    fun forDay(startOfDay: ZonedDateTime): ZonedDateTime? {
        return anyInRange(startOfDay.toEpochSecond(), startOfDay.plusDays(1).toEpochSecond())
    }

    fun anyInRange(start: Long, finish: Long): ZonedDateTime? {// FIXME: measure performance (weekdays especially)
        val zonedDateTime = getZonedDateTimeInRange(start, finish)
        val period = this.period
        if (zonedDateTime == null || exceptionRules.containsDate(zonedDateTime.toEpochDays()) ||
            (period is Period.Weekday && !period.verifyWeekday(// FIXME: works wacky for cases when more than one event in range
                getFirstZoneDateTime().toLocalDate(),
                zonedDateTime.toLocalDate()
            ))
        ) return null
        return zonedDateTime
    }

    fun getDesc(): String {
        return desc.ifEmpty { entry.name }
    }

    fun setDesc(desc: String) {
        this.desc = desc
    }

    fun makeEndless() {
        if (period is Period.Once)
            throw IllegalStateException("event without period cannot be endless")
        setTimesRepeat((END_ENDLESS_EXPECT - start) / period.secondsApproximation())
    }

    private fun ZonedDateTime.toEpochDays(): Long {
        return toLocalDate().toEpochDay()
    }

    private fun verifyParams(
        start: Long = this.start,
        timesRepeat: Long = this.timesRepeat,
        period: Period = this.period
    ) {
        if (timesRepeat < 0)
            throw IllegalArgumentException("timesRepeat must not be negative")
        try {
            if (timesRepeat > Period.MAX_PERIOD_COUNT ||
                period.secondsApproximation() > END_ENDLESS_THRESHOLD ||
                start + (timesRepeat - 1) * period.secondsApproximation() > END_ENDLESS_EXPECT
            )
                throw IllegalArgumentException("end overflow")
        } catch (e: ArithmeticException) {
            throw IllegalArgumentException("end overflow")
        }
    }

    fun getTimesRepeatUI(): Long {
        val period = this.period
        if (period is Period.Weekday) {
            val firstDay = getFirstZoneDateTime().toLocalDate()
            val wholePeriods = timesRepeat / (period.count * 7)
            val eventsPerWeek = period.eventCountPerWeek()
            var timesRepeatCumulative = wholePeriods * eventsPerWeek
            var leftoverDays = (timesRepeat % (period.count * 7)).toInt()
            if (leftoverDays == 0) {
                return timesRepeatCumulative
            } else {
                var then = firstDay.plusDays(period.count * 7)
                while (true) {
                    if (period.verifyWeekday(firstDay, then)) {
                        timesRepeatCumulative++
                    }
                    leftoverDays--
                    if (leftoverDays == 0) break
                    then = then.plusDays(1)
                }
                return timesRepeatCumulative
            }
        } else
            return timesRepeat
    }

    fun setTimesRepeatUI(timesRepeat: Long) {
        val period = this.period
        if (period is Period.Weekday) {
            val firstDay = getFirstZoneDateTime().toLocalDate()
            val eventsPerWeek = period.eventCountPerWeek()
            val eventsInFirstWeek =
                (period.daysMask shr (firstDay.dayOfWeek.value - 1)).countOneBits()
            var wholeWeeksCount = timesRepeat / eventsPerWeek * period.count
            var leftoverEvents = (timesRepeat % eventsPerWeek).toInt()
            if (eventsPerWeek == eventsInFirstWeek && leftoverEvents == 0)
                setTimesRepeat((wholeWeeksCount - period.count + 1) * 7 - firstDay.dayOfWeek.value + 1)
            else {
                if (leftoverEvents == 0) {
                    wholeWeeksCount -= period.count
                    leftoverEvents = eventsPerWeek
                }
                var then = firstDay.plusDays(wholeWeeksCount * 7)
                while (true) {
                    while (!period.verifyWeekday(firstDay, then)) then = then.plusDays(1)
                    if (--leftoverEvents == 0) break
                    then = then.plusDays(1)
                }
                setRange(startOfDayEnd = then.atStartOfDay(zoneId))
            }
        } else {
            setTimesRepeat(timesRepeat)
        }
    }

    private fun setTimesRepeat(timesRepeat: Long) {
        verifyParams(timesRepeat = timesRepeat)
        this.timesRepeat = timesRepeat
        this.end = getZoneDateTime(timesRepeat - 1).toEpochSecond()
    }

    fun setPeriod(period: Period) {
        if (isEndless) {
            this.period = period
            makeEndless()
            return
        }
        verifyParams(period = period)
        this.period = period
        this.end = getZoneDateTime(timesRepeat - 1).toEpochSecond()
    }

    fun addExceptions(dateStart: LocalDate, dateEnd: LocalDate = dateStart) {
        exceptionRules.addDates(dateStart.toEpochDay(), dateEnd.toEpochDay())
    }

    fun removeExceptions(dateStart: LocalDate, dateEnd: LocalDate = dateStart) {
        exceptionRules.removeDates(dateStart.toEpochDay(), dateEnd.toEpochDay())
    }

    fun setRange(startOfDayStart: ZonedDateTime? = null, startOfDayEnd: ZonedDateTime? = null) {
        if (startOfDayStart != null) {
            zoneId = startOfDayStart.zone
            val start = startOfDayStart.toEpochSecond()
            if (isEndless) {
                this.start = start
                makeEndless()
            } else {
                verifyParams(start = start)
                val end = getLastZoneDateTime()
                this.start = start
                if (period is Period.Once)
                    setTimesRepeat(1)
                else
                    setRange(startOfDayEnd = end)
            }
        }
        if (startOfDayEnd != null) {
            val startOfDayAfterEnd = startOfDayEnd.plusDays(1)
            val end = startOfDayAfterEnd.toEpochSecond() - 1
            var expectedCount = (end - start) / period.secondsApproximation() + 1
            verifyParams(timesRepeat = expectedCount)
            while (getZoneDateTime(expectedCount - 1) < startOfDayAfterEnd)
                expectedCount += (end - getZoneDateTime(expectedCount - 1).toEpochSecond()) / period.secondsApproximation() + 1
            while (getZoneDateTime(expectedCount - 1) >= startOfDayAfterEnd)
                expectedCount--
            if (expectedCount < 0)
                expectedCount = 0
            setTimesRepeat(expectedCount)
        }
    }

    @TestOnly
    fun getTimesRepeatForTesting() = timesRepeat

    @TestOnly
    fun getNextClosestForTesting(fromEpochSecond: Long) = getNextClosestRaw(fromEpochSecond)

    @TestOnly
    @Deprecated("For use only in bad testing data")
    fun fixEndForTesting(): Date {
        setTimesRepeat(timesRepeat)
        return this
    }

    @TestOnly
    fun getZDTForTesting(i: Long): ZonedDateTime {
        return getZoneDateTime(i)
    }

    fun fixExceptionList() { //fixme check period intersections
        if (timesRepeat <= 0)
            exceptionRules = ExceptionRules()
        exceptionRules.trimToFitRange(
            getZoneDateTime(0).toEpochDays(),
            getLastZoneDateTime().toEpochDays()
        )
    }

    private fun getNextClosestRaw(fromEpochSecond: Long): ZonedDateTime? {
        if (fromEpochSecond > end || timesRepeat <= 0)
            return null
        if (fromEpochSecond <= start)
            return getFirstZoneDateTime()
        if (period is Period.Weekday) {
            val period = this.period as Period.Weekday
            val firstLocal = getFirstZoneDateTime().toLocalDate()
            val weekNumberFirst = firstLocal.toWeekNumber()
            val start = Instant.ofEpochSecond(fromEpochSecond).atZone(zoneId)
            val startDay = start.toLocalDate()
            var then = startDay
            val weekNumberStart = then.toWeekNumber()
            val weeksDiff = weekNumberStart - weekNumberFirst
            if (weeksDiff % period.count == 0L) {//chosen start is in valid week
                if (period.verifyWeekday(firstLocal, then)) {//check that day
                    val thatDay = getZoneDateTime(then.toEpochDay() - firstLocal.toEpochDay())
                    if (fromEpochSecond <= thatDay.toEpochSecond()) {
                        return thatDay
                    }
                }
                //check until end of that week
                while (true) {
                    then = then.plusDays(1)
                    if (then.toWeekNumber() != weekNumberStart) break
                    if (period.verifyWeekday(firstLocal, then)) {
                        val validEvent = getZoneDateTime(then.toEpochDay() - firstLocal.toEpochDay())
                        return if (validEvent.toEpochSecond() > end) null else validEvent
                    }
                }
            }
            //set to next valid week
            then = firstLocal.plusWeeks((weeksDiff / period.count + 1) * period.count)
            //set then to start of week
            then = then.minusDays(then.dayOfWeek.value - 1L)
            //check until end of that week
            while (true) {
                if (period.verifyWeekday(firstLocal, then)) {
                    val validEvent = getZoneDateTime(then.toEpochDay() - firstLocal.toEpochDay())
                    return if (validEvent.toEpochSecond() > end) null else validEvent
                }
                then = then.plusDays(1)
            }
        }
        val period = (end - start) / (timesRepeat - 1)
        val idx = min((timesRepeat - 1), ((fromEpochSecond - start) / period))
        val zoneDateTime = getZoneDateTime(idx)
        val evt = zoneDateTime.toEpochSecond()
        if (evt < fromEpochSecond) {
            val zoneDateTimeP1 = getZoneDateTime(idx + 1)
            val evtNext = zoneDateTimeP1.toEpochSecond()
            return if (evtNext >= fromEpochSecond && idx < timesRepeat) zoneDateTimeP1 else null
        }
        if (idx == 0L) return getFirstZoneDateTime()
        val zoneDateTimeM1 = getZoneDateTime(idx - 1)
        val evtPrev = zoneDateTimeM1.toEpochSecond()
        return if (evtPrev < fromEpochSecond) zoneDateTime else zoneDateTimeM1
    }

    private fun getNextClosestIdxRaw(fromEpochSecond: Long): Long {
        if (fromEpochSecond > end || timesRepeat <= 0)
            return -1
        if (fromEpochSecond <= start)
            return 0
        val period = (end - start) / (timesRepeat - 1)
        val idx = min((timesRepeat - 1), ((fromEpochSecond - start) / period))
        val evt = getZoneDateTime(idx).toEpochSecond()
        if (evt < fromEpochSecond) {
            val evtNext = getZoneDateTime(idx + 1).toEpochSecond()
            return if (evtNext >= fromEpochSecond && idx < timesRepeat) idx + 1 else -1
        }
        if (idx == 0L) return 0
        val evtPrev = getZoneDateTime(idx - 1).toEpochSecond()
        return if (evtPrev < fromEpochSecond) idx else idx - 1
    }

    fun fixDateRange() {
        val epoch0 = ZonedDateTime.of(1970, 1, 1, 0, 0, 0, 0, zoneId)
        if (exceptionRules.containsDate(getLastZoneDateTime().toEpochDays())) {
            val endExceptionRange = exceptionRules.getRangeForDate(getLastZoneDateTime().toEpochDays())!!
            val newEnd = epoch0.plusDays(endExceptionRange.start - 1)
            if (newEnd.toEpochSecond() < start) {
                timesRepeat = 0
                end = start
            } else {
                setRange(startOfDayEnd = newEnd)
            }
        }
        if (exceptionRules.containsDate(getZoneDateTime(0).toEpochDays())) {
            val startExceptionRange = exceptionRules.getRangeForDate(getZoneDateTime(0).toEpochDays())!!
            if (timesRepeat != 0L) {
                val newStartIdx = getNextClosestIdxRaw(epoch0.plusDays(startExceptionRange.end + 1).toEpochSecond())
                if (newStartIdx != -1L)
                    setRange(startOfDayStart = getZoneDateTime(newStartIdx))
                else
                    setRange(startOfDayStart = getLastZoneDateTime())
            }
        }
    }

    companion object {
        const val TABLE_NAME: String = "dates"
        const val COLUMN_NAME_ID: String = "id"
        const val COLUMN_NAME_ENTRY_ID: String = "entryId"
        const val COLUMN_NAME_DESC: String = "description"
        const val COLUMN_NAME_TIME_START: String = "timeStart"
        const val COLUMN_NAME_DURATION: String = "duration"
        const val COLUMN_NAME_TIME_ENDS: String = "timeEnd"
        const val COLUMN_NAME_TIMES_REPEATS: String = "timesRepeat"
        const val COLUMN_NAME_PERIOD: String = "period"
        const val COLUMN_NAME_TIME_ZONE: String = "timeZone"
        const val COLUMN_NAME_REMOVED: String = "exceptionRules"
        const val END_ENDLESS_EXPECT: Long = 281474976710656L //2^48
        const val END_ENDLESS_THRESHOLD: Long = 140737488355328L //2^47

        fun cursorToList(
            dbManager: DbManager,
            cursor: Cursor
        ): MutableList<Date> {
            val dates: MutableList<Date> = ArrayList()
            val idIdx: Int = cursor.getColumnIndex(COLUMN_NAME_ID)
            val entryIdx: Int = cursor.getColumnIndex(COLUMN_NAME_ENTRY_ID)
            val descIdx: Int = cursor.getColumnIndex(COLUMN_NAME_DESC)
            val timeStartIdx: Int = cursor.getColumnIndex(COLUMN_NAME_TIME_START)
            val durationIdx: Int = cursor.getColumnIndex(COLUMN_NAME_DURATION)
            val timeEndsIdx: Int = cursor.getColumnIndex(COLUMN_NAME_TIME_ENDS)
            val timesRepeatsIdx: Int = cursor.getColumnIndex(COLUMN_NAME_TIMES_REPEATS)
            val periodIdx: Int = cursor.getColumnIndex(COLUMN_NAME_PERIOD)
            val timeZoneIdx: Int = cursor.getColumnIndex(COLUMN_NAME_TIME_ZONE)
            val removedIdx: Int = cursor.getColumnIndex(COLUMN_NAME_REMOVED)
            while (cursor.moveToNext())
                dates.add(
                    Date(
                        dbManager,
                        cursor.getInt(idIdx),
                        cursor.getInt(entryIdx),
                        cursor.getString(descIdx),
                        cursor.getLong(timeStartIdx),
                        cursor.getLong(durationIdx),
                        cursor.getLong(timeEndsIdx),
                        cursor.getLong(timesRepeatsIdx),
                        cursor.getString(periodIdx),
                        cursor.getString(timeZoneIdx),
                        cursor.getString(removedIdx)
                    )
                )
            return dates
        }

        fun getAllByEntryId(dbManager: DbManager, entryId: Int): List<Date> {
            dbManager.readableDatabase.rawQuery(
                "SELECT * FROM $TABLE_NAME WHERE $COLUMN_NAME_ENTRY_ID = ?",
                arrayOf(entryId.toString())
            ).use { cursor ->
                return cursorToList(dbManager, cursor)
            }
        }
    }
}